src/main/java/com/google/devtools/build/lib/skyframe/PrerequisiteProducer.java

```diff
@@ -318,6 +318,31 @@ public boolean evaluate(
     //  would exit this SkyFunction and restart it when permits were available.
     semaphoreLocker.acquireSemaphore();
     try {
+      // Check target compatibility before requesting toolchains - known missing toolchains are a
+      // valid reason for declaring a target incompatible and should not result in the target
+      // failing to build rather than being skipped as incompatible.
+      // Non-rule targets and those that are part of the toolchain resolution system do not support
+      // target compatibility checking anyway.
+      if (targetAndConfiguration.getTarget() instanceof Rule
+          && ((Rule) targetAndConfiguration.getTarget()).useToolchainResolution()) {
+        platformInfo = loadTargetPlatformInfo(env);
+        // loadTargetPlatformInfo may return null even when no deps are missing.
+        if (env.valuesMissing()) {
+          return false;
+        }
+
+        configConditions = computeConfigConditions(env, targetAndConfiguration, transitivePackages,
+            platformInfo, transitiveRootCauses);
+        if (configConditions == null) {
+          return false;
+        }
+
+        if (!checkForIncompatibleTarget(env, state, targetAndConfiguration, configConditions,
+            platformInfo, transitivePackages)) {
+          return false;
+        }
```

Comment:
How does this work for a `target_compatible_with` attribute that depends on the toolchain?  I can see that the tests are passing, but I don't understand how this is working in the code, apologies.

E.g. internally we have a few snippets like this:
```python
cc_library(
    ...
    target_compatible_with = select({
        # See if the `msan` feature is enabled in the toolchain.
        # If so, we can't build this target.
        ":msan_enabled": ["@platforms//:incompatible"],
        "//conditions:default": [],
    }),
)
```
How does this code allow for that? What am I missing? I suspect I'm overlooking something simple, my apologies.

Reply:
Hmm. Does `computeConfigConditions` load the toolchain if necessary?

Reply:
Could you provide the definition of `:msan_enabled`? I know that `select` can depend on the configuration as well as platform properties, which in turn influence toolchain resolution, but I don't know how to have a config setting that depends on properties of a particular toolchain directly.

Reply:
This is the .bzl code:
```python
load("@bazel_tools//tools/cpp:toolchain_utils.bzl", "use_cpp_toolchain")
load("@rules_cc//cc:toolchain_utils.bzl", "find_cpp_toolchain")

def get_cc_features(ctx):
    return cc_common.configure_features(
        ctx = ctx,
        cc_toolchain = find_cpp_toolchain(ctx),
        unsupported_features = ctx.disabled_features,
        requested_features = ctx.features,
    )

def _cc_feature_flag_impl(ctx):
    feature_configuration = get_cc_features(ctx)
    enabled = cc_common.is_enabled(
        feature_configuration = feature_configuration,
        feature_name = ctx.attr.feature,
    )
    return [config_common.FeatureFlagInfo(value = str(enabled))]

# Used to create select() statements based on a C++ toolchain feature.
cc_feature_flag = rule(
    implementation = _cc_feature_flag_impl,
    attrs = {
        "feature": attr.string(
            mandatory = True,
        ),
    },
    fragments = ["cpp"],
    toolchains = use_cpp_toolchain(),
    incompatible_use_toolchain_transition = True,
)
```
Then the build file looks like so:
```python
cc_feature_flag(
    name = "msan_enabled_flag",
    feature = "msan",
)

config_setting(
    name = "msan_enabled",
    flag_values = {
        ":msan_enabled_flag": "True",
    },
)
```
I think that's all of it.

Reply:
Or does this work because `msan_enabled` has a toolchain resolved for itself. Then the target with `target_compatible_with` doesn't need the toolchain resolved?

Reply:
@gregestren Please correct me if I'm wrong, but I think that's exactly what's happening here: `msan_enabled` has been analyzed already (which required resolving a C++ toolchain for it) by the time the `select` in `target_compatible_with` is evaluated.

Reply:
Should I hold off the internal review of this change? Are there more changes coming?

Reply:
@katre I think that you can proceed and won't make any further changes. I pushed two commits to resolve @gregestren's nit and fix a test failure that occurred in the state you originally approved.

Reply:
Yeah, I was confusing myself. I think @fmeum answered my question. Sorry for the noise!